File: lintdb/schema/Schema.cpp
#include "Schema.h"

namespace lintdb {
Json::Value Field::toJson() const {
    Json::Value json;
    json["name"] = name;
    json["data_type"] = static_cast<int>(data_type);

    Json::Value fieldTypesJson(Json::arrayValue);
    for (const auto& fieldType : field_types) {
        fieldTypesJson.append(static_cast<int>(fieldType));
    }
    json["field_types"] = fieldTypesJson;

    Json::Value params;
    params["dimensions"] = parameters.dimensions;
    params["analyzer"] = parameters.analyzer;
    params["quantization"] = static_cast<int>(parameters.quantization);
    params["num_centroids"] = parameters.num_centroids;
    params["num_iterations"] = parameters.num_iterations;
    params["num_subquantizers"] = parameters.num_subquantizers;
    params["nbits"] = parameters.nbits;
    json["parameters"] = params;

    return json;
}

Field Field::fromJson(const Json::Value& json) {
    Field field;
    field.name = json["name"].asString();
    field.data_type = static_cast<DataType>(json["data_type"].asInt());

    const Json::Value& fieldTypesJson = json["field_types"];
    for (const auto& fieldTypeJson : fieldTypesJson) {
        field.field_types.push_back(static_cast<FieldType>(fieldTypeJson.asInt()));
    }

    const Json::Value& params = json["parameters"];
    field.parameters.dimensions = params["dimensions"].asUInt();
    field.parameters.analyzer = params["analyzer"].asString();
    field.parameters.quantization = static_cast<QuantizerType>(params["quantization"].asInt());
    field.parameters.num_centroids = params["num_centroids"].asUInt();
    field.parameters.num_iterations = params["num_iterations"].asUInt();
    field.parameters.num_subquantizers = params["num_subquantizers"].asUInt();
    field.parameters.nbits = params["nbits"].asUInt();

    return field;
}

Json::Value Schema::toJson() const {
    Json::Value json;
    for (const auto& field : fields) {
        json["fields"].append(field.toJson());
    }
    return json;
}

Schema Schema::fromJson(const Json::Value& json) {
    Schema schema;
    for (const auto& jsonField : json["fields"]) {
        schema.fields.push_back(Field::fromJson(jsonField));
    }
    return schema;
}
} // namespace lintdb

File: lintdb/python/schema.py
from .lintdb import (
    Schema as LintDBSchema,
    FieldValue,
    Field as LintDBField,
    Schema as LintDBSchema,
    FieldParameters as LintDBFieldParameters,
    FieldTypeArray,
    FLOAT,
    INTEGER,
    TEXT,
    TENSOR,
    QUANTIZED_TENSOR,
    COLBERT,
    ColBERTContextData,
    FieldType_Stored as LintDBStored,
    FieldType_Indexed as LintDBIndexed,
    FieldType_Colbert as LintDBColBERT,
    IndexedField,
    StoredField,
    ContextField,
    ColbertField,
    BINARIZER,
    PRODUCT_ENCODER,
    NONE,
)
from typing import Dict, Any
from enum import Enum

class DataType(Enum):
    FLOAT = FLOAT
    INTEGER = INTEGER
    TEXT = TEXT
    TENSOR = TENSOR
    QUANTIZED = QUANTIZED_TENSOR
    COLBERT = ColBERTContextData

class FieldType(Enum):
    Stored = LintDBStored
    Indexed = LintDBIndexed
    Colbert = LintDBColBERT

Quantizer_Binarizer = BINARIZER
Quantizer_ProductEncoder = PRODUCT_ENCODER
Quantizer_None = NONE

def Stored(name: str, data_type: DataType, params=LintDBFieldParameters()):
    if isinstance(params, dict):
        params = FieldParameters(**params)

    return StoredField(name, data_type.value, params)

def Context(name: str, data_type: DataType, params=LintDBFieldParameters()):
    if isinstance(params, dict):
        params = FieldParameters(**params)

    return ContextField(name, data_type.value, params)

def Indexed(name: str, data_type: DataType, params=LintDBFieldParameters()):
    if isinstance(params, dict):
        params = FieldParameters(**params)

    return IndexedField(name, data_type.value, params)

def Colbert(name: str, data_type: DataType, params=LintDBFieldParameters()):
    if isinstance(params, dict):
        params = FieldParameters(**params)

    dt = data_type.value
    field = ColbertField(name, dt, params)
    field.add_field_type(LintDBColBERT)
    return field

def FieldParameters(**kwargs):
    """
        size_t dimensions = 0;
    std::string analyzer = "";
    QuantizerType quantization = QuantizerType::UNKNOWN;
    size_t num_centroids = 0;
    size_t num_iterations = 10;
    size_t num_subquantizers = 0; // used for PQ quantizer
    size_t nbits = 1; // used for PQ quantizer
    :param kwargs:
    :return:
    """
    result = LintDBFieldParameters()
    for key, value in kwargs.items():
        setattr(result, key, value)

    return result

def Schema(fields: [LintDBField]):
    result = LintDBSchema()
    for field in fields:
        result.add_field(field)

    return result

